src/io/read.rs

```diff
@@ -0,0 +1,41 @@
+use core::cmp;
+
+/// Non-blocking reader trait
+pub trait Read {
+    /// An enumeration of possible errors
+    ///
+    /// May be `!` (`never_type`) for infallible implementations
+    type Error;
+
+    /// Pull some bytes from this source into the specified buffer, returning how many bytes were
+    /// read.
+    ///
+    /// If an object needs to block for a read it will return an `Err(nb::Error::WouldBlock)`
+    /// return value.
+    ///
+    /// If the return value of this method is `Ok(n)`, then it must be guaranteed that `0 <= n <=
+    /// buf.len()`. The `n` value indicates that the buffer `buf` has been filled in with `n` bytes
+    /// of data from this source. If `n == 0 && buf.len() > 0` then it can be assumed that this
+    /// reader has run out of data and will not be able service any future read calls.
+    fn read(&mut self, buf: &mut [u8]) -> ::Result<usize, Self::Error>;
```

Comment:
I feel like I'm missing something, but shouldn't the return type here (and elsewhere) be `nb::Result<usize, Self::Error>`?

Is there a type alias somewhere that I'm missing?

Reply:
This is referring to `nb::Result`, it's just since this is defined in the `nb` crate the crate root is just `::`. This was a leftover from initially implementing in `embedded-hal` then doing minimal changes to get it compiling here, I'll switch to an actual `use` of `Result` here.

Reply:
Oh good lord that was silly of me. You linked from the `embedded-hal` crate, and I didn't notice that this was a pull request for `nb`